dephell_archive/_path.py

```diff
@@ -112,11 +112,20 @@ def iterdir(self, recursive: bool = False) -> Iterator['ArchivePath']:
             else:
                 members = descriptor.infolist()     # zip
 
+            top_level_items = set()
             # get files
             for member in members:
                 name = getattr(member, 'name', None) or member.filename
+                path, _sep, _name = name.partition('/')
+                if not recursive and path in top_level_items:
+                    continue
+
+                top_level_items.add(path)
                 yield self.copy(member_path=PurePath(name))
 
+            if not recursive:
+                 return
+
             # get dirs
             names = set()
```

Comment:
note that because tarfile and zipfile's return all items as a flat list, afaics, the rest of this code is not useful.

Reply:
I see why it is here; they can omit directories.

Reply:
Ideally all of the below is replaced with something like `_dir_list` from https://github.com/dephell/dephell_archive/pull/13 .  It is cached over there.  The cached list should be copy&filtered when creating new ArchivePath for subdirs.  But I think there are some bigger fishes to fry atm.